### PR TITLE
[Merged by Bors] - fix: do not use ToAdditive.applyAttributes outside ToAdditive

### PR DESCRIPTION
**Pull request by fpvandoorn**
* The `ToAdditive` version does some `to_additive`-specific things.
* Update docs to reflect that we apply the attribute to both the source and target declaration
* In Lean 3 the only attribute we ever copied was `simp`, which doesn't care if it's applied multiple times, so we don't check if the declaration already is applied to the source declaration. This is a risk, since it means that `@[simp, elementwise]` or `@[simp, reassoc]` will not be changed to `@[elementwise (attr := simp)]` or `@[reassoc (attr := simp)]`
* Note: copying the simp attribute from the source declaration is not a great solution, since the simp priority is not easily accessible in the environment (without walking the whole simp-`DiscrTree`)

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Util/AddRelatedDecl.lean [hunk 1]
```diff
@@ -66,6 +67,8 @@ def addRelatedDecl (src : Name) (suffix : String) (ref : Syntax)
...
   if isProtected (← getEnv) src then
     setEnv $ addProtected (← getEnv) tgt
-  if let some thisAttr := thisAttr then
-    let attrs := match attrs? with | some attrs => attrs | none => #[]
-    _ ← Term.TermElabM.run' <| ToAdditive.applyAttributes ref attrs thisAttr src tgt
+  let attrs := match attrs? with | some attrs => attrs | none => #[]
+  _ ← Term.TermElabM.run' <| do
+    let attrs ← elabAttrs attrs
+    Term.applyAttributes src attrs
+    Term.applyAttributes tgt attrs
```
Comment (kmill) [inline Mathlib/Util/AddRelatedDecl.lean:74]: Untested:
```suggestion
  _ ← Term.TermElabM.run' <| do
    let attrs ← elabAttrs (attrs?.getD #[])
    Term.applyAttributes src attrs
    Term.applyAttributes tgt attrs
```

**Review by kmill — APPROVED** [review]
Looks good to me.

@semorrison Do you have any thoughts on how this affects `elementwise`/`reassoc`? The change means that the translation of `@[simp, elementwise, reassoc]` to Lean 4 is `@[elementwise (attr := simp), reassoc (attr := simp)]`. As Floris explained to me, there's not a good way to copy the `simp` attribute from one definition to another anymore.

Something we might look into is defining an attribute that adds attributes to every declaration that another attribute introduces, so for example `@[attrs_and_apply [elementwise, reassoc] (attr := simp)]`, where the `attr` attributes are applied to everything including the current declaration (I'm assuming you'd do something like `whatsnew` to see what to apply to). Maybe that's overengineering things though.

**Comment by fpvandoorn** [comment]
Note: `simp` also sometimes auto-generates new declarations, so beware of that (I don't think that is necessarily a problem with your suggestion).
`@[to_additive (attr := simp) addDecl] def mulDecl` already does something similar (but a bit more complicated): if `simp` on `mulDecl` generated `mulAuxSimpLemma` and `simp` on `addDecl` generated `addAuxSimpLemma`, then we also add `@[to_additive addAuxSimpLemma]` to `mulAuxSimpLemma`

File: Mathlib/Util/AddRelatedDecl.lean (hunk 1 above)
Comment (fpvandoorn) [reply Mathlib/Util/AddRelatedDecl.lean:74]: There is actually a coercion on line 70, so this doesn't work

**Comment by semorrison** [comment]
> The change means that the translation of `@[simp, elementwise, reassoc]` to Lean 4 is `@[elementwise (attr := simp), reassoc (attr := simp)]`.

I'm still a bit surprised by this syntax. I would have expected this to put `simp` on the declarations generated by `@[elementwise]` and `@[reassoc]`, but not on the original declaration!

**Comment by fpvandoorn** [comment]
I can imagine. The thing is that with `to_additive`, `elementwise` and `reassoc`, 90+% of the time you want the `simp` attribute (or another attribute) on the original lemma iff you want it on the new lemma, so this saves a bunch of keystrokes.

**Comment by semorrison** [comment]
Not sure why CI isn't running here.

**Comment by Parcly-Taxel** [comment]
> Not sure why CI isn't running here.

It didn't run on #3132 when GitHub itself was experiencing an incident…

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**